### ISSUE TITLE
As a new user discovering Composer, my path is clearly laid out

### ISSUE THREAD

**Issue by EdMoffatt** [hyperledger/composer#912]
Community members have expressed confusion with how they're meant to get started with Composer: what should they install first?  Do they need a Fabric running before they set up the CLI?  etc.  We should:

1. Agree the expected/ideal route for a user to take through discovering Composer
2. Update the experience to make this route the path of least resistence
3. Also be careful to allow for users deciding to skip over steps if they're interested in a particular thing (e.g. "don't care about web playground, show me the VSCode plugins")

I'll add an initial idea of this expected route to kick us off (in a comment), and then create some child work items as we identify the tasks needed to achieve this.

**Comment by EdMoffatt** [comment]
Proposed "golden path" for a totally new user discovering Hyperledger Composer:

1. Try it out before you install anything: visit the web-hosted Playground.  Here you can import and edit samples, or create your own business network from scratch.  You can also test how it would work if deployed, by creating assets, participants and transactions that are stored in browser memory.

2. Now you've decided you like Composer - excellent!  Next step is to get a local  dev environment running: local Playground pointed at a local Fabric.  We'll get you the latest versions of both, fired up via docker-compose, and we'll launch the Composer Playground on your localhost, already configured to deploy business networks to your Fabric.

3. At some stage, you might want to stop using Playground and use IDE plugins + CLI: here’s how!

At this point, you're fully set up with a Hyperledger Composer dev environment.  Afterwards, there are a couple of likely directions you could want to go in:

4. Now you want to deploy to a Fabric that’s not running locally?  No problem!  Here’s how…

5. Now you want to generate APIs and integrate with other things (like a front end app you're developing for example)?  No problem!  Here's how...

**Comment by EdMoffatt** [comment]
Suggested work in support of this (to be split out into issues if we agree they're needed!):

- Rework the first few sections of the doc to promote this path
- Establish whether "latest versions of both" (Composer and Fabric) - which means "the latest versions that play well together for sure and we have tested that" - can be achieved in such a way that we don't have to update the commands if newer versions come out
- Change the docker-compose setup so that the Composer Playground instance is using the hlfabric connection profile by default when first started
- Any rework necessary to "Quickstart" so that it becomes the CLI-setup tutorial.  I'd suggest we have less focus on describing the Digital Property Network and doing transactions with it.
- Make sure we have a good section on "deploying elsewhere" (i.e. a solid intro to connection profiles)
- Make sure we have a good section on "APIs, apps, integration"

**Comment by seanbarclay** [comment]
deploying somewhere is great. but the thing to delineate is deploy to what? a fabric simulation environment or a production based network w/participants running their own infra. 

it might be helpful to come up with best practices around them. integrations could be for individual participants or could be shared. that yields interesting opportunities as well.

**Comment by dselman** [comment]
I agree this should be a priority. @mbwhite is working on cleaning up the Fabric Dev Setup scripts and moving them out of Getting Started. I would like to promote this sort of path (all up for discussion):

0. Use Playground on the web
1. Install Playground + Fabric Local Dev Server -> connection profile pre-created
2. Install Composer Local Dev Tools (CLI, REST Server, Angular Generator)
3. Install VSCode/Atom plugin (optional)
4. Tutorial 1...N

Let's get rid of the confusing Quick Start vs. Getting Started, and instead have a well organised set of description tutorials. @EdProsser what do you think?

**Comment by EdMoffatt** [comment]
That suggestion is excellent, @dselman - a really strong flow imo.

Will raise this on the community call on Thursday too, see if others have thoughts and feedback on the approach before we start the re-jig!

**Comment by EdProsser** [comment]
> Let's get rid of the confusing Quick Start vs. Getting Started, and instead have a well organised set of descriptive tutorials:

These will go away regardless, as part of the docs review work. Was always the plan to nuke them. 

>Exploring the Playground
Create a new Business Network Definition
Create a command line application
Expose a REST API
Create an Angular application
Deploy a Business Network Archive to HSBN
Deploy a Node.js client application to Bluemix
Deploy a Node.js client application to AWS

This is fine, but I think it's worth bearing in mind that different personas will want to use the path differently. Will devs want to model the business network in playground _before_ creating the command line application? The ordering of creating a new business network definition and then installing local tools is the only question I have.

**Comment by trauts2** [comment]
I just started investigating Composer, and I love the theme of this issue. 

Regarding the REST tutorial, it would be nice to show how a custom API method would be added.  Or if the framework tooling requires sticking strictly to REST conventions for nouns and verbs with no RPC-like methods, that could be stated there as well.  This would help calibrate new devs to the current capabilities and limitations of the current composer framework.

**Comment by EdMoffatt** [comment]
We discussed this on yesterday's community call, and this is the diagram I drew out to talk through it:

![image](https://cloud.githubusercontent.com/assets/7454304/25991563/7e21507c-36fb-11e7-9528-18dca4401ab7.png)

(Just pasting in here for people who weren't on the call to see!)

**Comment by EdProsser** [comment]
Here's my sizings for the DOCS portion of work from Mark's email:

>As a new user, I want to try Composer Playground with no local installation, so that I can familiarise myself with its concepts and capabilities.

There's little to no docs work here. Giving this a 1.
>As a new user, I want to install and run Composer Playground and Fabric locally, so that I can experiment offline with persistent state. (This could be broken into smaller, incremental stories).

There's some docs work here, but we've got the body of content, just needs correctly adapting and arranging. Giving this a 5.
>As a developer, I want to install Composer development tools locally, so that I have an environment in which to develop Composer applications.

These docs I've not spent too much time looking at, so giving this an 8 to cover the worst case scenarios.
>As a developer, I want to learn how to use the Composer CLI, so that I can develop and deploy Composer applications.

These docs are actually alright, so giving this like a 3?
>As a developer, I want to use the REST API to Composer, so I can create a REST client application.

This one's a decent amount of work. 8.
>As a developer, I want to generate an Angular UI for my business network, so that I can demonstrate my application to the business.

We have docs, but not great ones. 5.
>As a developer, I want to deploy my business network to HSBN.

Haven't documented anything on this note. 8.
>As a developer, I want to deploy my Node.js client application to Bluemix.

Is deploying an application to an external service something we want to document? Other than: Here's the docs that service has written. 
>As a developer, I want to deploy my Node.js client application to AWS.

Same as HSBN nothing yet. 8.

Maybe worth nothing, that these aren't sizes for the job as a whole, this is purely the time investment it'll require from me.

**Comment by EdProsser** [comment]
BTW those sizings are written in the spirit of Planning Poker.

**Comment by EdMoffatt** [comment]
1.  As a new user, I want to try Composer Playground with no local installation, so that I can familiarise myself with its concepts and capabilities.

> This one seems to purely be a case of making sure we pull any changes made through other issues into the web Playground instructions.  I could even be convinced that the doc could just say "go have a play, and come back here when you want to get it installed locally!"

My Sizing: 1

2.  As a new user, I want to install and run Composer Playground and Fabric locally, so that I can experiment offline with persistent state. (This could be broken into smaller, incremental stories).

> This is a case of adjusting things in line with Fabric releases.  It's doc and it's also updates to the scripts that do the local install.  We have a separate issue for the "one liner install script" which we'd love to be able to offer as the method in this documentation bit too.  There's also the issue to make the hlfabric profile be the one we default to instead of the web one with the local install, which will mean that the user has arrived at this point successfully a little faster.

My Sizing: 1

3.  As a developer, I want to install Composer development tools locally, so that I have an environment in which to develop Composer applications.

> The current experience of doing this is fine, but then it's only included as part of the Quickstart right now, which is a bit obtuse.  I think we'd need to pull this out separately in the doc, explain what the tools are for, and add a quick way of testing they're working for a user who's already running Playground and Fabric locally (e.g. could we use them to get some info from one of those containers to prove that it's working?).  I'm not sure what Design work there is to do here, sounds more like doc and some dev work to make sure we have a way of testing that the install worked without Quickstart?

My Sizing: 3

4.  As a developer, I want to learn how to use the Composer CLI, so that I can develop and deploy Composer applications.

> This is covered by the docs currently known as Quickstart and Getting Started.  There are definitely v1 support things to do with Getting Started.  So there's probably a fairly sizable piece of work to do here.  We need to form a single tutorial, we need to have a single sample network that's used for it.

My Sizing: 5

5.  As a developer, I want to use the REST API to Composer, so I can create a REST client application.

> I've never successfully got through this part of the documentation.  Therefore I'm convinced there's a fair bit of work to do here ;)

My Sizing: 5

I don't feel I can contribute much to the other sizings yet: I haven't managed to make it through these bits of the docs, so I really need to get to grips with their current state before I can say how much work I think is there!

**Comment by mahoney1** [comment]
Adding my two cents (euros or US).. 

1) the thing that's missing for me is 'a tour' (however its done) or 'tour experience' is perhaps more succinct. . I think it would help address the issue of  'my [the user's] path is clearly laid out'.  So 1) the user can 'start the tour', navigate to step 1 (playground) and so on along the path (as @dselman suggested),  but flexible to see any step 'explored' from a nav perspective and 2) provide a tour of that particular step selected 3) allow a return to start,  or prev, next in sequence etc.  The goal of the tour is to give a 'big picture' experience of what's on offer and give the user his/her 'bearings'. Doesn't detract from tutorials or docs etc described here needed anyway. . Shepherd would be useful for tours I believe. Anyway, worth consideration, if it helps the user experience.

2) Add the end-to-end Digital Property sample to the list - existing 'consumers' of this, find this useful - we have an 0.6 and 1.0 tutorial and includes the 'cause and effect' change (ie of app logic) as part of that.

_Exploring the Playground
Create a new Business Network Definition
Create a command line application_
Create and amend an end-to-end sample digital property CLI application.
_Expose a REST API
Create an Angular application
Deploy a Business Network Archive to HSBN
Deploy a Node.js client application to Bluemix
Deploy a Node.js client application to AWS__

Note that a simple tour could be done for each Tutorial for a consistent experience. 

Happy to go off and create a Shepherd sample.

**Comment by mahoney1** [comment]
hi Ed, just circling back (after assignment) - I have a sample 'tour' of a Composer page set up, in an Ubuntu VM, which I can demo/explain/discuss use case. Its not Shepherd as it turns out, something better, if we can use.

**Comment by sstone1** [comment]
Have folded #1107 into this epic as this area will be completely rewritten.

**Comment by mbwhite** [comment]
One thing that stands out to be... "what's the difference between an APPLICATION and a BUSINESS-NETWORK??"

The difference, implicitly,  get explained in the later sections. I think you'd pick it up from the end of the Developer tutorial

Early on (somewhere around Ed's banana diagram) I think we need to put applications into context as well. Possibly use the phrase Application SDK.
-----

The Developer Tutorial feels too long to be honest - esp if there's nothing at the top to set my expectation of how long this is.  Should it be Network Developer Tutorial, and Application SDK Tutorial?

**Comment by mbwhite** [comment]
https://hyperledger.github.io/composer/unstable/getting-started/using-playground-locally.html
Download link should be..

    curl -O https://hyperledger.github.io/composer/install-hlfv1.sh

**Comment by caroline-church** [comment]
# https://hyperledger.github.io/composer/unstable/introduction/key-concepts.html  
- [ ] On this page it says ```Hyperledger Composer can be experienced through a browser-based UI called Hyperledger Composer Playground. This can be used via a hosted version (no install necessary) or a local install (good for editing and testing sample Business Networks offline).``` But it is possible to test your business network and edit it on the hosted version you just can't use a real blockchain. 
- [ ] One of the key concepts is listed as `Blockchain State Storage` is this the right title?

# https://hyperledger.github.io/composer/unstable/getting-started/getting-started.html
- [ ] `or get setup up with development tools for working on the command line straight away.` do we need the `straight away` here?

# https://hyperledger.github.io/composer/unstable/getting-started/getting-started-with-playground.html
- [ ] `A GitHub account (if you want to import sample business network from GitHub)` should this be either `a sample business network` or `sample business networks`
- [ ] `clearing your browser storage` this should be `browser local storage`

# https://hyperledger.github.io/composer/unstable/getting-started/development-tools.html
- [ ] `If you've installed the Hyperledger Composer Playground locally you'll need to close the containers by using the following scripts.` Is `closing the containers` the right terminology to use here those too commands removes the containers
- [ ] `For v1.0-alpha, there is nothing to as this the default ` this should be `For v1.0-alpha, there is nothing to do....`
- [ ] `If you want to swap between v0.6 and v1.0, ensure you have issued a stopFabric.sh and a teardownFabric.sh command first be START on the other version` the last part of this sentance doesn't make sense

**Comment by sstone1** [comment]
https://hyperledger.github.io/composer/unstable/getting-started/using-playground-locally.html

1. curl -O https://hyperledger.github.io/composer/composer/install-hlfv1.sh
2. cat install-hlfv1.sh | bash

These should be a single step:

1. curl -sSL https://hyperledger.github.io/composer/composer/install-hlfv1.sh | bash

It's not single click if you have to run two commands!

**Comment by sstone1** [comment]
It's also a bit weird that this appears _before_ you get to the install:

> Please note: If you've previously used Hyperledger Composer Playground locally or Hyperledger Fabric and wish to clear everything and start again, the following commands will delete any running containers and delete all downloaded images, (be careful if you're using other Docker images on your machine):

The one click install should take care of downloading the latest images and removing any conflicting Docker images. I would move it afterwards into a "How to clean up" type of section.

**Comment by jt-nti** [comment]
@EdProsser Playground links should always be https. There seem to be some http links in getting-started-with-playground.md

**Comment by sstone1** [comment]
https://hyperledger.github.io/composer/unstable/getting-started/development-tools.html

Remove all references to `composer-playground-api`. You don't need to install it separately.

**Comment by mbwhite** [comment]
_Introduction_ - How does Hyperledger Composer work in **_practice_**
Should there be a line underneath the Where do I go from here?

_Typical Solution Architecture_
Diagram consuses the issue about if the SDK can or can't be used by applications. Also intimates that the SDK has to connected to by and only by REST

**Comment by sstone1** [comment]
Need a +999 emoji for @jt-nti comment.

**Comment by jt-nti** [comment]
Looks like the title is wrong in getting-started-with-playground.md
> title: Tutorial 1 - Building your first business network
..and probably the excerpt.

**Comment by mbwhite** [comment]
Diagnosing problems needs a rework  re the logging.  I'll get the info for that post lunch

**Comment by sstone1** [comment]
Features and Concepts in Hyperledger Composer

This list of concepts should really mention Registry and maybe Identity as well.

This section is wrong:

> Blockchain State Storage

> The current state of assets and participants are stored on the blockchain (ledger), for example, as administered by Hyperledger Fabric. Hyperledger Fabric distributes the ledger across a set of peers and ensures that updates to the ledger are consistent across all peers (consensus).

I would replace "Blockchain State Storage" with ... something else. 

The paragraph needs to be reworded as well. Transactions (inputs) get recorded on the ledger, the current state of assets and participants are stored in the state database. These are two separate things within a Blockchain. My rewording:

All transactions submitted to the business network are stored on the Blockchain ledger, and the current state of assets and participants are stored in the Blockchain state database, for example, as administered by Hyperledger Fabric. Hyperledger Fabric distributes the ledger and the state database across a set of peers and ensures that updates to the ledger and state database are consistent across all peers using a consensus algorithm.

**Comment by sstone1** [comment]
https://hyperledger.github.io/composer/unstable/introduction/key-concepts.html

Diagram has Rest, which should be REST.
Angular 2, when it should be Angular 4 - let's just drop the version number and call it Angular instead, unless we really want to regenerate all of the text and diagrams every time we upgrade it.

https://hyperledger.github.io/composer/unstable/introduction/solution-architecture.html

Loopback needs to be LoopBack
Angular 2 needs to be Angular

I won't go looking for them all but it's probably in other places as well.

**Comment by sstone1** [comment]
Some of our screenshots seem to be huge, at least on Safari:

https://hyperledger.github.io/composer/unstable/getting-started/getting-started-with-playground.html
https://hyperledger.github.io/composer/unstable/tutorials/developer-guide.html

**Comment by sstone1** [comment]
The REST server screen shot is out of date on this page (and other pages):

https://hyperledger.github.io/composer/unstable/tutorials/developer-guide.html

**Comment by sstone1** [comment]
On the very front page: https://hyperledger.github.io/composer/unstable/index.html
We have:

> Calls happen every Thursday:

The trailing colon should be replaced with a full stop.

**Comment by sstone1** [comment]
https://hyperledger.github.io/composer/unstable/applications/genapp.html

We should change the title to capitalise the N in Node.js.

**Comment by mahoney1** [comment]
hi @sstone1 , on 'big image' comment  and other image sizes: for  https://hyperledger.github.io/composer/unstable/tutorials/developer-guide.html . I have updated (eg, v1,0 changes, Hyperledger Composer' notation, etc - also I have replaced the REST Server screenshot) and I've committed the changes to the guide just now.  Have reduced some images, may need to reduce one or two more.

**Comment by dselman** [comment]
We should remove the instruction to `npm install -g` the composer-rest-server in the Developer Guide -- that was already done in the "Developer Tools Install"
https://hyperledger.github.io/composer/unstable/tutorials/developer-guide.html

**Comment by jt-nti** [comment]
Minor point but this seems like an odd thing to focus on at the top of _Installing and running Hyperledger Composer Playground locally_ using-playground-locally.md

> Hyperledger Composer Playground can also be used in a "browser only" mode, without a running instance of Hyperledger Fabric. When used in this mode, all the functionality of Hyperledger Composer Playground is available, but all of the data (business networks, assets, participants, and transactions) is persisted into browser local storage.

With a few edits, that seems more appropriate for the running online topic...

> The Composer Playground is a tool which provides a simple UI for quickly editing and testing a business network.
The hosted version (Online Playground) can only be used in a "browser only" mode, without a running instance of Hyperledger Fabric. When used in this mode, all of the data (business networks, assets, participants, and transactions) is persisted into browser local storage.
Alternatively you can install the Playground locally (Local Playground) when you want to connect to a running instance of Hyperledger Fabric. Skip to the Install Local Playground instructions if you'd like to start using a Local Playground.
Note: There is no install necessary to use the Online Playground.

Leaving something more like this for the Install Local Playground instructions...

> This tutorial will take you through how to install and run the Hyperledger Composer Playground on your local machine. It also creates an instance of Hyperledger Fabric v1.0-Alpha1.
Note: There is no need to install a Local Playground if you do not intend to connect to a running instance of Hyperledger Fabric. Skip to Try Hyperledger Composer Playground online instructions to use the Online Playground

Just a thought.

**Comment by sstone1** [comment]
Something about "Install Development Tools" doesn't sound right to me.
Installing a local Fabric is not something I'd call a "Tool".
I think "Install Development Environment" would be a better heading.

**Comment by mbwhite** [comment]
- [ ] Hover text for the top links is inconsistent and not useful on the front page only
- [ ] Should the playground images be link to something?  eg just bring it up bigger... 
- [ ] Boxes in the community page aren't the same side.

**Comment by jt-nti** [comment]
- [ ] We should be using **example.org** instead of the domain name of a _Central New Jersey networking organization_.

(I'm looking at "Using the Playground to run a Car Auction" but we do this kind of thing all over the place, e.g. **org.acme** elsewhere.)

**Comment by mahoney1** [comment]
@dselman  - thanks - have removed the `npm install -g composer-rest-server` from https://hyperledger.github.io/composer/unstable/tutorials/developer-guide.html and ensure it refers to link to development-tools.md (as pre-requisite steps to set up Dev env etc) -  waiting for PR to merge all changes to the developer-guide.md

**Comment by EdProsser** [comment]
All above change (that are docs text related) have been made.

**Comment by dselman** [comment]
@EdProsser @EdMoffatt I get the impression we should close this now. Can you please check nothing was missed and open new issues as necessary?

**Comment by EdMoffatt** [comment]
I think this one will be ready to close with @EdProsser 's next PR :)

I'm going to raise separate issues for a couple of hanging things, namely:

- Work out what we're going to do about a "tour" for first time Playground users
- Get some love on the Tutorials, generally tidying up and making sure we have good screenshots in there!

**Comment by sstone1** [comment]
Boom, closed.
